src/validate/validate_state_property.ts

```diff
@@ -0,0 +1,214 @@
+import {ValidationError} from '../error/validation_error';
+import {getType} from '../util/get_type';
+import {deepUnbundle, unbundle} from '../util/unbundle_jsonlint';
+import {isObjectLiteral} from '../util/is_object_literal';
+
+interface ValidateStatePropertyOptions {
+    key: string;
+    value: unknown;
+}
+
+export function validateStateProperty({
+    key,
+    value,
+}: ValidateStatePropertyOptions): ValidationError[] {
+    if (!isObjectLiteral(value)) {
+        return [
+            new ValidationError(
+                key,
+                value,
+                `object expected, ${getType(value)} found`
+            ),
+        ];
+    }
+    if (value.type === undefined && value.enum === undefined) {
+        return [
+            new ValidationError(
+                key,
+                value,
+                'schema must have a "type" or "enum" property'
+            ),
+        ];
+    }
+
+    switch (unbundle(value.type)) {
+        case 'string':
+            return validateString(value, key);
+        case 'number':
+            return validateNumber(value, key);
+        case 'boolean':
+            return validateBoolean(value, key);
+        case 'array': {
+            return validateArray(value, key);
+        }
+        default: {
+            if (value.type !== undefined) {
+                return [
+                    new ValidationError(
+                        'type',
+                        value.type,
+                        'expected string, number, boolean or array'
+                    ),
+                ];
+            }
+        }
+    }
+
+    return validateEnum(value, key);
+}
+
+function validateString(schema: Record<string, unknown>, key: string) {
+    const defaultValue = unbundle(schema.default) as string;
```

Comment:
Is `unbundle` needed in all these places?

Reply:
typeof checks fail otherwise. I think it's safer to cast to primitive types explicitly and` if type is already primitive, unbundle` internally will just return it:

![image](https://github.com/user-attachments/assets/9eaddb35-40cf-431a-928e-f87829e5df4e)